x-pack/platform/plugins/shared/lens/public/visualizations/tagcloud/tagcloud_visualization.tsx

```diff
@@ -104,16 +110,22 @@ export const getTagcloudVisualization = ({
 
   triggers: [VIS_EVENT_TO_TRIGGER.filter],
 
-  initialize(addNewLayer, state, mainPalette) {
-    return (
-      state || {
-        layerId: addNewLayer(),
-        layerType: LayerTypes.DATA,
-        ...DEFAULT_STATE,
-        colorMapping:
-          mainPalette?.type === 'colorMapping' ? mainPalette.value : getColorMappingDefaults(),
-      }
-    );
+  initialize(addNewLayer, state, mainPalette, datasourceState) {
+    if (state) return convertToRuntimeState(state, datasourceState);
+
+    return {
+      layerId: addNewLayer(),
+      layerType: LayerTypes.DATA,
+      ...DEFAULT_STATE,
+      colorMapping:
+        mainPalette?.type === 'colorMapping' ? mainPalette.value : getColorMappingDefaults(),
+    };
+  },
+
+  isEqual(state1, references1, datasourceState1, state2, references2, datasourceState2) {
+    const convertedState1 = convertToRuntimeState(state1, datasourceState1);
+    const convertedState2 = convertToRuntimeState(state2, datasourceState2);
+    return deepEqual(convertedState1, convertedState2);
```

Comment:
This implementation (and the other ones except for XY) doesn't consider the `references` in the equality, is it safe to skip that? I'm not so sure

Reply:
Yeah, only the xy chart currently uses the `references`, I thought about making a conditional type definition as is done on the `initialize` method but decided to just keep it.

They all share the same implementation so there are things that one vis needs that others don't. Could definitely use a bit of cleanup.